Eliminate unsafe buffer usage in GraphicsContextGLCVCocoa and GraphicsContextCG

geoffreygaren opened the pull request:
#### da8462128955b0a095f803704003b777f12582da
<pre>
Eliminate unsafe buffer usage in GraphicsContextGLCVCocoa and GraphicsContextCG
<a href="https://bugs.webkit.org/show_bug.cgi?id=286100">https://bugs.webkit.org/show_bug.cgi?id=286100</a>
<a href="https://rdar.apple.com/143082904">rdar://143082904</a>

Reviewed by Chris Dumez.

I pre-merged a little code from <a href="https://github.com/WebKit/WebKit/pull/39148">https://github.com/WebKit/WebKit/pull/39148</a> so
we don&apos;t end up conflicting.

* Source/WTF/wtf/IndexedRange.h:
(WTF::BoundsCheckedIterator::begin):
(WTF::BoundsCheckedIterator::end):
(WTF::BoundsCheckedIterator::BoundsCheckedIterator):
(WTF::boundsCheckedBegin):
(WTF::boundsCheckedEnd): use std::begin / std::end so we can do indexed and
zipped ranged for loops on C arrays.

* Source/WTF/wtf/TrailingArray.h: Paragraph separate first from last. (I came
here to try to add data(), but it&apos;s there already.)

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageImpl): Updated for span.

* Source/WebCore/platform/graphics/GraphicsContextGLImageExtractor.h:
(WebCore::GraphicsContextGLImageExtractor::imagePixelData): Use span and
MallocSpan because UniqueArray doesn&apos;t know its length.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::setCGStyle): Use zippedRange instead of raw
indexing into a C array.

* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::getSourceDataFormat): Use std::array and std::span.

(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.mm: Use std::array.

Canonical link: <a href="https://commits.webkit.org/289065@main">https://commits.webkit.org/289065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d0b7b42f4e413e0e6d6a5030937869578759b2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90319 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36232 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87281 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12902 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66241 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24049 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77377 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46510 "Found 2 new API test failures: /WPE/TestSSL:/webkit/WebKitWebView/web-socket-client-side-certificate, /WPE/TestSSL:/webkit/WebKitWebView/web-socket-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3683 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31637 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35300 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78160 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32469 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91726 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84233 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12539 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9126 "Found 1 new test failure: http/tests/navigation/page-cache-requestAnimationFrame.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74746 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12765 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73214 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73864 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18286 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16712 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4526 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13283 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12482 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17931 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106623 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12312 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25710 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14062 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -154,8 +154,12 @@ class FixedVector {
     T& operator[](size_t i) LIFETIME_BOUND { return m_storage->at(i); }
     const T& operator[](size_t i) const LIFETIME_BOUND { return m_storage->at(i); }
 
+    T* data() LIFETIME_BOUND { return m_storage->data(); }
```
But we removed those to encourage people to use `span()` iirc. And you could always use `span().data()` if you *really* need a raw pointer.

geoffreygaren replied:
Hmmmmm... OK.

cdumez commented on Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
```diff
@@ -109,189 +108,189 @@ uint8_t convertColor16BigTo8(uint16_t value)
         destination[3] = convertColor16LittleTo8(source[0]);
-        source += 1;
-        destination += 4;
+        source = source.subspan(1);
```
`skip(source, 1);` from `<wtf/text/ParsingUtilities.h>`

cdumez commented on Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
```diff
@@ -109,189 +108,189 @@ uint8_t convertColor16BigTo8(uint16_t value)
-        source += 1;
-        destination += 4;
+        source = source.subspan(1);
+        destination = destination.subspan(4);
```
`skip(destination, 4);`

cdumez replied:
This comment seems to apply to other places as well.

cdumez commented on Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
```diff
@@ -72,22 +71,22 @@ enum AlphaFormat {
-        { GraphicsContextGL::DataFormat::BGR8,  GraphicsContextGL::DataFormat::RGB8,  GraphicsContextGL::DataFormat::RGB16Little,  GraphicsContextGL::DataFormat::RGB16Big },
-        { GraphicsContextGL::DataFormat::ABGR8, GraphicsContextGL::DataFormat::RGBA8, GraphicsContextGL::DataFormat::RGBA16Little, GraphicsContextGL::DataFormat::RGBA16Big },
-        { GraphicsContextGL::DataFormat::BGRA8, GraphicsContextGL::DataFormat::ARGB8, GraphicsContextGL::DataFormat::ARGB16Little, GraphicsContextGL::DataFormat::ARGB16Big }
+    static const auto formatTable = std::array { // SourceDataFormat::Base x bitsPerComponent x endian
```
Why not:
```
static std::array formatTable {
```

It is more concise.

cdumez replied:
Applies elsewhere in this PR too.

cdumez reviewed (approved):
LGTM

geoffreygaren committed:
Eliminate unsafe buffer usage in GraphicsContextGLCVCocoa and GraphicsContextCG
https://bugs.webkit.org/show_bug.cgi?id=286100
rdar://143082904

Reviewed by Chris Dumez.

I pre-merged a little code from https://github.com/WebKit/WebKit/pull/39148 so
we don't end up conflicting.

* Source/WTF/wtf/IndexedRange.h:
(WTF::BoundsCheckedIterator::begin):
(WTF::BoundsCheckedIterator::end):
(WTF::BoundsCheckedIterator::BoundsCheckedIterator):
(WTF::boundsCheckedBegin):
(WTF::boundsCheckedEnd): use std::begin / std::end so we can do indexed and
zipped ranged for loops on C arrays.

* Source/WTF/wtf/TrailingArray.h: Paragraph separate first from last. (I came
here to try to add data(), but it's there already.)

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageImpl): Updated for span.

* Source/WebCore/platform/graphics/GraphicsContextGLImageExtractor.h:
(WebCore::GraphicsContextGLImageExtractor::imagePixelData): Use span and
MallocSpan because UniqueArray doesn't know its length.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::setCGStyle): Use zippedRange instead of raw
indexing into a C array.

* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::getSourceDataFormat): Use std::array and std::span.

(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cv/GraphicsContextGLCVCocoa.mm: Use std::array.

Canonical link: https://commits.webkit.org/289065@main
